pallets/reward-distribution/src/lib.rs

```diff
@@ -94,10 +169,229 @@ pub mod pallet {
+
+			//withdraw the reward for specific nominator
+			let caller_rewards =
+				ext::staking::withdraw_reward::<T>(&vault_id, &caller, index, reward_currency_id)?;
+
+			if caller_rewards == (BalanceOf::<T>::zero()) {
+				return Err(Error::<T>::NoRewardsForAccount.into())
+			}
+
+			//transfer rewards
+			Self::transfer_reward(reward_currency_id, caller_rewards, caller)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn transfer_reward(
+		reward_currency_id: CurrencyId<T>,
+		reward: BalanceOf<T>,
+		beneficiary: T::AccountId,
+	) -> DispatchResult {
+		let native_currency_id = T::GetNativeCurrencyId::get();
+		if reward_currency_id == native_currency_id {
+			let available_native_funds = T::Balances::free_balance(&Self::fee_pool_account_id());
+
+			match reward.checked_sub(&available_native_funds) {
+				None => {
+					// Pay the whole reward from the fee pool
+					let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+					return amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+				},
+				Some(remaining) => {
+					//check if the to-be-minted amount is consistent
+					//with reward tracking
+					let liability = NativeLiability::<T>::get()
+						.ok_or(Error::<T>::NotEnoughRewardsRegistered)?;
+
+					if liability < remaining {
+						return Err(Error::<T>::NotEnoughRewardsRegistered.into())
+					}
+
+					// Use the available funds from the fee pool
+					let available_amount: currency::Amount<T> =
+						Amount::new(available_native_funds, reward_currency_id);
+					available_amount.transfer(&Self::fee_pool_account_id(), &beneficiary)?;
+					// Mint the rest
+					T::Balances::deposit_creating(&beneficiary, remaining);
+
+					NativeLiability::<T>::set(Some(
+						liability.checked_sub(&remaining).ok_or(Error::<T>::Underflow)?,
+					));
+				},
+			}
+			Ok(())
+		} else {
+			//we need no checking of available funds, since the fee pool will ALWAYS have enough
+			// collected fees of the wrapped currencies
+			let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+			amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+		}
+	}
+
+	pub fn execute_on_init(_height: T::BlockNumber) {
+		if let Err(_) = ext::security::ensure_parachain_status_running::<T>() {
+			return
+		}
+
+		//get reward per block
+		let reward_per_block = match RewardPerBlock::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("Reward per block is None");
+				return
+			},
+		};
+
+		let rewards_adapted_at = match RewardsAdaptedAt::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("RewardsAdaptedAt is None");
+				return
+			},
+		};
+
+		let mut reward_this_block = reward_per_block;
+
+		let decay_interval_sub = match T::DecayInterval::get().checked_sub(&T::BlockNumber::one()) {
+			Some(value) => value,
+			None => {
+				log::warn!("Decay interval underflow");
+				return
+			},
+		};
+
+		if let Ok(expired) =
```

Comment:
With the usage of `checked_sub` this ends up being even more convoluted. Taking into account [that differentiation](https://github.com/pendulum-chain/spacewalk/pull/415#discussion_r1373245872) which I think we should keep let me know @ebma if you can think of a way to simplify this.

Reply:
Let's just use a [saturating_sub](https://snowbridge-rust-docs.snowfork.com/sp_runtime/traits/trait.saturating#tymethod.saturating_sub) then as this does not throw an error, thus we do not need to add a complex match expression.

Reply:
I think you could just do 
```
if Ok(true) == ext::security::parachain_block_expired::<T>(
			rewards_adapted_at,
			T::DecayInterval::get().saturating_sub(T::BlockNumber::one()),
		) {
			let decay_rate = T::DecayRate::get();
			...
...
```